File: app/src/main/java/edu/upenn/aradha/starling/droidStar/lp/AsyncTaskLP.java
package edu.upenn.aradha.starling.droidStar.lp;

import java.util.List;
import java.util.ArrayList;

import java.lang.Thread;
import java.util.concurrent.TimeUnit;

import android.os.AsyncTask;
import android.os.Handler.Callback;
import android.content.Context;


import edu.colorado.plv.droidStar.LearningPurpose;
import static edu.colorado.plv.droidStar.Static.*;

public class AsyncTaskLP extends LearningPurpose {
    protected SimpleTask task;
    public int counter;

    // INPUTS
    public static String EXECUTE = "exec";
    public static String CANCEL = "cancel";
    public static String GET = "get";

    public static String PARAM = "asdf";

    protected List<String> uniqueInputSet() {
        List<String> inputs = new ArrayList();
        inputs.add(EXECUTE);
        inputs.add(CANCEL);
        inputs.add(GET);

        return inputs;
    }

    // OUTPUTS
    public static String CANCELLED = "cancelled";
    public static String POSTEXEC = "postexec";
    public static String PREEXEC = "preexec";

    public boolean isError(String output) {
        // there are no errors for this class?
        return false;
    }

    public int betaTimeout() {
        return 500;
    }

    // public int safetyTimeout() {
    //     // Maybe this will fix the onCancelled issues
    //     return 2000;
    // }

    public String shortName() {
        return "AsyncTask";
    }

    // public int postResetTimeout() {
    //     // The task could take up to 200ms to complete, so we need to
    //     // wait longer than that!!!
    //     return 500;
    // };

    public AsyncTaskLP(Context c) {
        super(c);
        this.task = null;
        this.counter = 0;
    }

    protected void resetActions(Context context, Callback callback) {
        doReset();
    }

    protected void doReset() {
        if (task != null) {
            task.cancel(true);
            counter++;
        }
        task = new SimpleTask(counter);
    }

    public void giveInput(String input) throws Exception {
        if (input.equals(EXECUTE)) {
            task.execute(PARAM);
        } else if (input.equals(CANCEL)) {
            task.cancel(false);
        } else if (input.equals(GET)) {
            task.get(0, TimeUnit.MILLISECONDS);
        } else {
            logl("Unknown command to AsyncTask");
            throw new IllegalArgumentException("Unknown command to AsyncTask");
        }
    }

    public class SimpleTask extends AsyncTask<String, String, String> {
        int localCounter;

        public SimpleTask(int c) {
            super();
            this.localCounter = c;
        }
        protected String doInBackground(String... ss) {
            try {Thread.sleep(200);}
            catch (Exception e) {logl("Sleep problem?");}
            return PARAM;
        }

        protected void onPreExecute() {
            // respond(PREEXEC);
        }

        protected void onCancelled(String s) {
            if (localCounter == counter) {
                respond(CANCELLED);
            }
        }

        protected void onPostExecute(String s) {
            if (s.equals(PARAM)) {
                respond(POSTEXEC);
            }
        }
    }

}

File: src/main/scala/edu/colorado/plv/droidstar/experiments/lp/AsyncTaskLP.scala
package edu.colorado.plv.droidstar
package experiments.lp

import android.content.Context
import android.os.Handler.Callback

import android.os.AsyncTask


import scala.collection.JavaConverters._

class AsyncTaskLP(c: Context) extends LearningPurpose(c) {
  var task: AsyncTask[AnyRef,AnyRef,AnyRef] = null
  var counter: Int = 0

  val param = "asdf"

  val execute = "exec"
  val cancel = "cancel"
  val cancelled = "on_cancelled"
  val postexec = "on_postexec"

  override def betaTimeout(): Int = 500

  @throws(classOf[Exception])
  override def giveInput(i: String, altKey: Int): Unit = i match {
    case `execute` => task.execute(param)
    case `cancel` => task.cancel(false)
    case _ => {
      logl("Unknown command to AsyncTask")
      throw new IllegalArgumentException("Unknown command to AsyncTask")
    }
  }

  override def inputAlts(): java.util.Map[String,java.lang.Integer] = {
    val m = super.inputAlts()
    m.put(execute,1)
    m
  }

  override def isError(o: String): Boolean = false

  override def resetActions(c: Context, b: Callback): String = {
    if (task != null) {
      task.cancel(true)
      counter += 1
    }
    task = new SimpleTask(counter)
    null
  }
  override def shortName(): String = "AsyncTask"
  override def uniqueInputSet(): java.util.List[String] =
    List(execute,cancel).asJava

  class SimpleTask(localCounter: Int) extends AsyncTask[AnyRef,AnyRef,AnyRef] {

    /* Using `String` instead of `AnyRef` gives AbstractMethodErrors...
     *
     * (see https://stackoverflow.com/questions/24934022/asynctask-doinbackground-abstract-method-not-implemented-error-in-android-scal)
     */
    override def doInBackground(ss: AnyRef*): AnyRef = {
      try {Thread.sleep(200)}
      catch {
        case _ : Throwable => logl("Sleep problem?")
      }
      param
    }
    override def onCancelled(s: AnyRef): Unit = {
      if (localCounter == counter) {
        respond(cancelled)
      }
    }
    override def onPostExecute(s: AnyRef): Unit = {
      respond(postexec)
    }
  }

}
